### PR TITLE
SOF-1858 - Create Configurable Static Buttons Card

### PR DESCRIPTION
**Pull request by RasmusAlbrektsen**
To allow configurable static buttons a new component has been created for this specific purpose using a `staticActionIds` property on the `ShelfConfiguration`. 

Requires the Sofie Server PR: https://github.com/tv2/sofie-server/pull/121

Note to reviewers:
I'm very open to design opinions so if you have any improvements then let's discuss!
Also, this breaks if you have a prior shelfConfiguration. So you need to change the `staticActionIds` from null to undefined in the database. Either update via Postman or directly manipulate the value in the database.

### CODE REVIEW

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.scss [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      background-color: var(--light-gray-color);
+      border: solid 3px var(--gray-color);
+
+      overflow-y: auto;
+      overflow-x: hidden;
+    }
+
+    .button-group {
+      width: 100%;
+      margin-top: 25px;
```
Comment (vivanov90) [inline src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.scss]: You can use `_sofie-utility-classes` for spacings

File: src/app/settings/shelf-settings/components/shelf-action-panel-settings-page/shelf-action-panel-settings-page.component.ts
Comment (vivanov90) [inline src/app/settings/shelf-settings/components/shelf-action-panel-settings-page/shelf-action-panel-settings-page.component.ts:1]: In this file don't have any real changes

File: src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      this.updateStaticActions()
+    })
+    this.actionStateService
+      .subscribeToSystemActions()
+      .then(observable => {
+        observable.subscribe(actions => (this.availableActions = actions as Tv2Action[]))
+        this.updateStaticActions()
+      })
+      .catch(error => this.logger.data(error).error('Error while listening to Action events'))
+  }
```
Comment (vivanov90) [inline src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.ts]: You have 3 subscribes and you can use takeUntil and destroySubject can unsubscribe on destroy for all of them

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<div class="edit-static-buttons-menu">
+    <div class="selected-actions">
+        <ng-container *ngFor="let action of updatedSelectedStaticActions">
+            <sofie-action-box class="p-2" [action]="action" (click)="removeAction(action)"></sofie-action-box>
+        </ng-container>
```
Comment (KvelaGorrrrnio) [inline src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html]: The `ng-container` can be removed if the `*ngFor` is moved to the `sofie-action-box` element.

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    this.updatedSelectedStaticActions.push(this.selectedAction)
+  }
+
+  public removeAction(action: Tv2Action): void {
+    const index: number = this.updatedSelectedStaticActions.indexOf(action)
+    if (index < 0) {
+      return
+    }
+    this.updatedSelectedStaticActions.splice(index, 1)
```
Comment (KvelaGorrrrnio) [inline src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.ts]: An alternative can also be to use a filter like so:
```typescript
this.updatedSelectedStaticActions = updatedSelectedStaticActions.filter(action => action.id !== actionToRemove.id)
```
Note that I use `id` as the equality base.

File: src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss [hunk 5]
```diff
@@ -0,0 +1,18 @@
+:host {
+  display: flex;
+
+  .card-header {
+    font-size: 1.5em;
```
Comment (KvelaGorrrrnio) [inline src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss]: Should this be `1.5rem`?

File: src/app/settings/shelf-settings/components/edit-static-buttons-dialog/edit-static-buttons-dialog.component.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<sofie-sidebar-container i18n-title title="edit-static-buttons-dialog.header">
```
Comment (LindvedKrvang) [inline src/app/settings/shelf-settings/components/edit-static-buttons-dialog/edit-static-buttons-dialog.component.html]: We should change `title` to something else.
`i18n-title` adds a tooltip almost everywhere inside the component...
image

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html (hunk 3 above)
Comment (LindvedKrvang) [inline src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html]: You don't need the `ng-container` here. You can just put the `*ngFor` directly on `sofie-action-box`

**Review by LindvedKrvang — COMMENTED** [review]
Overall the code looks code.
There are some unfortunate behaviour with the design, but we have talked about those. Waiting with approval until they have been fixed :)

File: src/app/settings/shelf-settings/components/edit-static-buttons-dialog/edit-static-buttons-dialog.component.html (hunk 6 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/edit-static-buttons-dialog/edit-static-buttons-dialog.component.html]: I have changed it to follow way it's done in other files. This fixes the issue you mention!

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html (hunk 3 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html]: Thanks for that! It has been changed!

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.scss (hunk 1 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.scss]: I have changed it to instead be more than we have put as a maximum. I don't know if there a limits to how many utility classes we want to have before it doesn't make sense anymore.

File: src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.ts (hunk 2 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.ts]: Changed so all have unsubscribes with two using the destroySubject approach.

**Review by vivanov90 — APPROVED** [review]

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html (hunk 3 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.html]: Was removed in earlier change!

File: src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.ts (hunk 4 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/edit-static-buttons/edit-static-buttons.component.ts]: Very simple and nice, i like it! Changed!

File: src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss (hunk 5 above)
Comment (RasmusAlbrektsen) [reply src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss]: I have changed it, i think that makes more sense yeah :-D Though it is not my strongest area of expertise.

File: src/app/settings/shelf-settings/components/edit-static-buttons-dialog/edit-static-buttons-dialog.component.html (hunk 6 above)
Comment (LindvedKrvang) [reply src/app/settings/shelf-settings/components/edit-static-buttons-dialog/edit-static-buttons-dialog.component.html]: What I meant for you to do was to change the _attribute_ `title` on `sofie-sidebar-container` to something else than title like we did on `sofie-page`.

File: src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.html [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        <sofie-icon-button [icon]="Icon.PEN" [iconSize]="IconSize.M" [tooltip]="'Edit Static Buttons'" (click)="openEditStaticButtons()"></sofie-icon-button>
+    </ng-container>
+        <div class="static-buttons-display">
+            <ng-container *ngIf="selectedStaticActions.length !== 0 else noStaticButtons">
+                <div class="button-grid">
+                    <ng-container *ngFor="let action of selectedStaticActions">
+                        <sofie-action-box [action]="action"></sofie-action-box>
+                    </ng-container>
+                </div>
+            </ng-container>
```
Comment (LindvedKrvang) [inline src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.html]: You don't need these ng-containers, you can put the directives directly on the markup

File: src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  padding-left: 10px;
+}
+
+.static-buttons-display {
+  margin-left: 30px;
+  padding-bottom: 20px;
+  overflow-y: auto;
+  overflow-x: hidden;
+
+  .button-grid {
```
Comment (LindvedKrvang) [inline src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss]: Is this still a grid? ;)

Comment (LindvedKrvang) [reply src/app/settings/shelf-settings/components/static-buttons-configuration-card/static-buttons-configuration-card.component.scss]: And it's technically not buttons ;)

**Review by LindvedKrvang — APPROVED** [review]
Looks good.

You have a tendency to still set margins and paddings through CSS which we would like to minimize.

**Pull request merged**